### PR TITLE
Implement an algorithm for photon/pi0 discrimination based on input features and a trained model in ONNX format

### PR DESCRIPTION
**Pull request by giovannimarchiori**
The code in this PR provides a Gaudi algorithm that 
- reads the list of input features from a JSON file and the trained BDT model from an ONNX file
- for each cluster, reads the input features from the shapeParameters vector (if available), runs the inference, and saves the photon probability as an additional shape parameter of the new output cluster collection
It works with ONNX files created with either XGBoost and LightGBM (tested with both).
Could probably work also with different models based on features (DNN..) 

Uses as inputs the shower shapes calculated in https://github.com/HEP-FCC/k4RecCalorimeter/pull/79

### CODE REVIEW

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  /// Handle for output calorimeter clusters collection
+  mutable DataHandle<edm4hep::ClusterCollection> m_outClusters {
+    "outClusters", Gaudi::DataHandle::Writer, this
+  };
+
+  /// Handles for the cluster shower shape metadata to read and to write
+  MetaDataHandle<std::vector<std::string>> m_inShapeParameterHandle{
+    m_inClusters,
+    edm4hep::shapeParameterNames,
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.h:88]: Can you use `edm4hep::labels::ShapeParameterNames` instead?

Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.h]: done

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+  return outClusters;
+}
+
+StatusCode PhotonIDTool::readMVAFiles(const std::string& mvaInputsFileName,
+				      const std::string& mvaModelFileName)
+{
+  // 1. read the file with the list of input features
+  // TODO: figure how to save BDT to ONNX such that input is
+  // the list of input features with their names, rather
+  // than single tensor "X"
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:164]: Can you use `nlohmann_json` instead of parsing the JSON manually? It is a dependency anyway (make it explicit in the CMake file). To keep ability to have comments:
https://json.nlohmann.me/features/comments/

Also, why are you splitting part of the configuration outside of the steering file?
Do you have that many shape parameters?

If this is just temporary solution, ignore this comment.

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+      // otherwise, we will use the energy of the cluster object
+      auto it = std::find(shapeParameters.begin(), shapeParameters.end(), "rawE");
+      if (it != shapeParameters.end())
+	{
+	  int position = std::distance(shapeParameters.begin(), it);
+	  m_inputPositionsInShapeParameters.push_back(position);
+	  info() << "Feature " << feature << " found in position " << position << " of shapeParameters" << endmsg;
+	}
+      else {
+	m_inputPositionsInShapeParameters.push_back(-1);
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:71]: `m_inputPositionsInShapeParameters` is `unsigned short int`

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 2 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:164]: Hi @kjvbrt, the number of shape parameters can exceed 100.
What are you referring to in particular when you write "splitting part of the configuration outside of the steering file"? Do you mean, putting the names of the shape variables in the json rather than in the .py that steers k4run? Or something else?
About nlohmann_json, I thought it was an overkill for just parsing a json that contains a list of strings (could have been a CSV). But OK, in the end I can probably use that, so that in the json I will also save the parameters used at training time which is always useful to record somewhere

Comment (kjvbrt) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:164]: > What are you referring to in particular when you write "splitting part of the configuration outside of the steering file"? Do you mean, putting the names of the shape variables in the json rather than in the .py that steers k4run? Or something else?

Yes, this is what I mean, I would like to keep all configurations needed for running in one file, if possible. I agree that 100 parameters is a bit too much for the k4run steering file :)

> About nlohmann_json, I thought it was an overkill for just parsing a json that contains a list of strings (could have been a CSV). But OK, in the end I can probably use that, so that in the json I will also save the parameters used at training time which is always useful to record somewhere

I would not suggest using it if it was not already a hidden dependency.

What is your TODO about, is it that you might be able read the list of the shape parameters directly from the ONNX file?

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    // pass data through model
+    try
+    {
+      std::vector<Ort::Value> output_tensors = m_ortSession->Run(m_input_names,
+                                                                 input_tensors,
+                                                                 m_output_names,
+                                                                 Ort::RunOptions{nullptr});
+
+      // double-check the dimensions of the output tensors
+      // NOTE: the number of output tensors is equal to the number of output nodes specifed in the Run() call
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:331]: specifed -> specified

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+	  int position = std::distance(shapeParameters.begin(), it);
+	  m_inputPositionsInShapeParameters.push_back(position);
+	  info() << "Feature " << feature << " found in position " << position << " of shapeParameters" << endmsg;
+	}
+      else {
+	m_inputPositionsInShapeParameters.push_back(-1);
+      }
+    }
+    else {
+      // for the other features, check if they are in the shape paramaters
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:75]: paramaters -> parameters

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 2 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:164]: yes, the TODO was about that, but it seems not possible unfortunately

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 5 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp]: fixed

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 4 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp]: fixed

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 3 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp]: changed to vector of short int

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp (hunk 2 above)
Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp]: Hi @kjvbrt I implemented the parsing of the json with nlohmann. The json stores all relevant info from the training of the photonID algorithm and is generated automatically by the tool that trains the BDT and creates the onnx model, so using it rather than copying to the k4run steering file in my opinion provides some better consistency.
I saw that you merged some PRs so I will probably have to rebase my PR. Apart from that is there anything else I should change? Thanks a lot!

File: RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+  // Open the JSON file
+  std::ifstream file(mvaInputsFileName);
+  if (!file.is_open()) {
+    error() << "Error opening file: " << mvaInputsFileName << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // Parse the JSON file
+  json j;
+  file >> j;
```
Comment (kjvbrt) [inline RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:176]: What happens if the file contains non valid JSON?

Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp:176]: failing miserably I guess, need to probably put this in some try .. catch block. On my todo for next week

Comment (giovannimarchiori) [reply RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp]: fixed

**Comment by giovannimarchiori** [comment]
Hi @kjvbrt , I implemented the handling of exceptions in reading the JSON and rebased - is there anything left for me to do?

Thx, Giovanni

**Comment by kjvbrt** [comment]
Hi @giovannimarchiori, nice :)
Merging...
